examples/multigpu_advtrain/trainer.py

```diff
@@ -0,0 +1,211 @@
+
+
+def update_learning_params(trainer, manager):
+    model = manager.model
+    hparams = manager.hparams
+    fd = trainer.feed_dict
+    trainer.step_num = trainer.step_num
+
+    if hparams.model_type == 'resnet_tf':
+        if trainer.step_num < hparams.lrn_step:
+            lrn_rate = hparams.resnet_lrn
+        elif trainer.step_num < 30000:
+            lrn_rate = hparams.resnet_lrn/10
+        elif trainer.step_num < 35000:
+            lrn_rate = hparams.resnet_lrn/100
+        else:
+            lrn_rate = hparams.resnet_lrn/1000
+
+        fd[model.lrn_rate] = lrn_rate
+
+
+def create_train_graph_multigpu(manager):
+    assert '_multigpu' in manager.hparams.attack_type_train
+
+    hparams = manager.hparams
+    model = manager.model
+    x_pre, x, y = manager.g0_inputs
+    sess = manager.sess
+    batch_size = manager.batch_size
+
+    # Generates steps on gpus 0-(ngpu-1)
+    logging.info("Initializing train attack %s" % hparams.attack_type_train)
+    inputs, outputs = get_attack(model, x, hparams.attack_type_train, sess,
+                                 batch_size=batch_size, y=y,
+                                 nb_iter=hparams.attack_nb_iter_train,
+                                 dataset=hparams.dataset, ngpu=hparams.ngpu)
+
+    assert len(inputs) == len(outputs)
+    # 0
+    # inputs[0] = (x_pre, y)
+
+    # copy y forward
+    for i in range(len(outputs)):
+        if i > 0:
+            with tf.device(inputs[i][-1].device):
+                y2 = clone_variable('y%d' % i, y)
+        else:
+            y2 = y
+        inputs[i] = inputs[i] + (y2,)
+        outputs[i] = outputs[i] + (y2,)
+
+    # train step on last gpu
+    x, adv_x, y = outputs[-1]
+    device_name = '/gpu:%d' % (hparams.ngpu-1)
+    model.set_device(device_name)
+    with tf.device(device_name):
+        with tf.variable_scope('last'):
+            x2 = clone_variable('x_-1', x)
+            adv2_x = clone_variable('adv_x_-1', adv_x)
+            y2 = clone_variable('y_-1', y)
+            inputs += [(x2, adv2_x, y2)]
+            if not hparams.adv_train:
+                preds = model.fprop(x2, training=True, bn_training=True)
+                preds_2_adv = None
+            elif not hparams.only_adv_train:
+                preds = model.fprop(x2, training=True)
+                preds_2_adv = model.fprop(adv2_x, training=True,
+                                          bn_training=True)
+            else:
+                preds = None
+                preds_2_adv = model.fprop(adv2_x, training=True,
+                                          bn_training=True)
+            train_fetches = build_train_op(manager, preds, y2, preds_2_adv)
+
+    outputs += [train_fetches]
+
+    device_name = '/gpu:%d' % (hparams.ngpu-1)
+    model.set_device(device_name)
+    with tf.device(device_name):
+        sync_ops = model.create_sync_ops(host_device=device_name)
+
+    return inputs, outputs, sync_ops
+
+
+def model_train(trainer, manager):
```

Comment:
there's more than one `model_train` in this PR, can you comment on the differences between them?

Reply:
This is now part of TrainManager and TrainerMultiGPU and TrainerSingleGPU with most of the code being shared. The refactoring is done in commits between 4bd545ec41930128bfbc397fd34699e3c9e2c0e2 and f75fc0d7d4c361685578685f302a522bbd4fb2fa.